src/field255.rs

```diff
@@ -0,0 +1,308 @@
+};
+use fiat_crypto::curve25519_64::{
+    fiat_25519_add, fiat_25519_carry, fiat_25519_carry_mul, fiat_25519_from_bytes, fiat_25519_opp,
+    fiat_25519_sub, fiat_25519_tight_field_element, fiat_25519_to_bytes,
+};
+use num_bigint::{BigInt, Sign};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    convert::TryFrom,
+    fmt::{self, Debug, Display, Formatter},
+    hash::{Hash, Hasher},
+    io::{Cursor, Read},
+    marker::PhantomData,
+    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+};
+
+static MODULUS_BIG_ENDIAN: [u8; 32] = [
+    0x7f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xed,
+];
+
+/// `GF(2^255 - 19)`, a 255-bit field.
+#[derive(Clone, Copy, Default)]
+pub struct Field255(fiat_25519_tight_field_element);
+
+/// Attempts to instantiate a `Field255` from the first `Self::ENCODED_SIZE` bytes in the
+/// provided slice.
+///
+/// # Errors
+///
+/// An error is returned if the provided slice is not long enough to encode a field
+/// element or if the decoded value is greater than the field prime.
+impl Field255 {
+    fn try_from_bytes(bytes: &[u8], mask_top_bit: bool) -> Result<Self, FieldError> {
+        if Self::ENCODED_SIZE > bytes.len() {
+            return Err(FieldError::ShortRead);
+        }
+
+        let mut value = [0u8; Self::ENCODED_SIZE];
+        value.copy_from_slice(&bytes[..Self::ENCODED_SIZE]);
+
+        if mask_top_bit {
+            value[0] &= 0x7f;
+        }
+
+        let mut less_than_modulus = false;
+        for (value_byte, modulus_byte) in value.iter().zip(MODULUS_BIG_ENDIAN.iter()) {
+            match value_byte.cmp(modulus_byte) {
+                std::cmp::Ordering::Less => {
+                    less_than_modulus = true;
+                    break;
+                }
+                std::cmp::Ordering::Equal => continue,
+                std::cmp::Ordering::Greater => return Err(FieldError::ModulusOverflow),
+            }
+        }
+        if !less_than_modulus {
+            return Err(FieldError::ModulusOverflow);
+        }
+
+        // Switch from big endian to little endian.
+        value.reverse();
+        let mut output = [0; 5];
+        fiat_25519_from_bytes(&mut output, &value);
+
+        Ok(Field255(output))
+    }
+}
+
+impl PartialEq for Field255 {
+    fn eq(&self, rhs: &Self) -> bool {
+        // The internal representation used by fiat-crypto is not 1-1 with the field, so it is
+        // necessary to compare field elements via their canonical encodings.
+
+        let mut self_encoded = [0; 32];
+        fiat_25519_to_bytes(&mut self_encoded, &self.0);
+        let mut rhs_encoded = [0; 32];
+        fiat_25519_to_bytes(&mut rhs_encoded, &rhs.0);
+
+        self_encoded == rhs_encoded
+    }
+}
+
+impl Eq for Field255 {}
+
+impl Hash for Field255 {
```

Comment:
It looks like we don't use this internally. (but it would allow using `HashSet<Field255>` or `HashMap<Field255, _>`) I had written this because the other field types do the same, but the trait does not actually require a `Hash` bound.

Reply:
Maybe Janus needs it somewhere...

Reply:
I tried patching in `prio` with the hash implementations in `field.rs` deleted, and both `prio-server` and Janus work fine without that. Dropping `impl Hash for Field255` should be OK.

Reply:
In the interest of keeping PRs minimal, I suggest dropping it.